src/referral_maintenance.rs

```diff
@@ -22,25 +36,149 @@ pub async fn maintenaince_tasks(db: Arc<ReferralStore>, dune_data_folder: String
         {
             let mut guard = match db.0.lock() {
                 Ok(guard) => guard,
-                Err(poisoned) => poisoned.into_inner(),
-            };
-            match guard.app_data.get(&app_data) {
-                Some(_) => {}
-                None => {
-                    guard.app_data.insert(app_data, None);
-                }
+                Err(_) => return Err(anyhow!("Mutex poisoned")),
             };
+            guard
+                .app_data
+                .entry(app_data)
+                .or_insert(Referral::TryToFetchXTimes(3));
         }
     }
+    // 2st step: get all unintialized referrals
+    let uninitialized_app_data_hashes: Vec<H256> = {
+        let guard = match db.0.lock() {
+            Ok(guard) => guard,
+            Err(_) => return Err(anyhow!("Mutex poisoned")),
+        };
+        guard
+            .app_data
+            .clone()
+            .into_iter()
+            .filter(|(_, referral)| matches!(referral, Referral::TryToFetchXTimes(_)))
+            .map(|(hash, _)| hash)
```

Comment:
One minor performance win would be to:
```suggestion
            .iter()
            .filter(|(_, referral)| matches!(referral, Referral::TryToFetchXTimes(_)))
            .map(|(hash, _)| hash)
            .copied()
```

This way, you only copy the individual hashes instead of copying the whole `HashMap` first.